### ISSUE TITLE
ptu file format support?

### ISSUE THREAD

**Issue by snizzleorg** [Photon-HDF5/phconvert#12]
Is there any plans to support the more recent ptu-file format from PicoQuant

**Comment by tritemio** [comment]
We are willing to add support for any file format that is used in the field. If you want to submit a patch to add support for PTU files I can help you with the process. 

To support a new file format you need to add a "reader" function. In this case the file is from PicoQuant so the function should be added in `phconvert/pqreader.py`. 

# Current implementation

Currently, `pqreader` decodes HT3 files in 3 steps:

1. Function [ht3_reader()](https://github.com/Photon-HDF5/phconvert/blob/0.6.9/phconvert/pqreader.py#L39): decodes the HT3 header and gets the raw t3 records
2. Function [ht3_process()](https://github.com/Photon-HDF5/phconvert/blob/0.6.9/phconvert/pqreader.py#L158): extracts timestamps, detector numbers and TCSPC nanotimes from raw t3 records
3. Function [_correct_overflow()](https://github.com/Photon-HDF5/phconvert/blob/0.6.9/phconvert/pqreader.py#L189): performs overflow correction on timestamps. Since this step is relatively slow there are two versions of the function: one is pure python and the other uses numba (which is orders of magnitude faster and is used by default when numba is installed).

The user only calls the high-level function [pqreader.load_ht3()](https://github.com/Photon-HDF5/phconvert/blob/master/phconvert/pqreader.py#L20) which returns the decoded timestamps, detectors, nanotime and metadata.

# What you need to do

Most likely you just need to write a new function for step 1 (e.g. `ptu_reader()`) and reuse the existing functions for step 2 and 3.

Finally, we can rename `pqreader.load_ht3()` to `pqreader.load_picoquant()` and make it handle all picoquant formats.

If you decide to do it, you should also send a sample PTU file (ideally with PIE or nsALEX data) so I can upload it to [figshare](https://figshare.com/articles/Examples_data_files_for_phconvert/1455963). We have a [testing notebook](https://github.com/Photon-HDF5/phconvert/blob/master/tests/Execute%20notebooks.ipynb) performing all the conversions automatically to catch regressions. I'll need to add PTU to the list of tested formats and update the docs here and there.

Let me know if you need more info.

**Comment by snizzleorg** [comment]
Thanks for the explanation. If I find some time I will try to look into it. Congratulation on the great software anyways.

**Comment by tritemio** [comment]
Thanks @snizzleorg. BTW, I added this explanation to the phconvert docs:

http://phconvert.readthedocs.org/en/latest/pqreader.html

**Comment by tritemio** [comment]
Just in case you need it, there is a guide on using git & github for contributing to phconvert here #13.

**Comment by tritemio** [comment]
Guide for contributing through GitHub now on the wiki: 

https://github.com/Photon-HDF5/phconvert/wiki/How-to-contribute-on-Github

**Comment by tritemio** [comment]
@snizzleorg, this message is to let you know that @talaurence implemented support for pt3 decoding in phconvert. Now, with the two examples of ht3 and pt3, adding ptu support should be even easier.

**Comment by snizzleorg** [comment]
thanks for letting me know. I am following this... However I'm not sure if I can make time to implement it nonetheless if somebody is willing I can however supply *.ptu files

**Comment by tritemio** [comment]
@snizzleorg, ok no problem. Can you upload a *.ptu data file somewhere and send a link here? Maybe somebody will come and implement it ;). What measurements do you have? nsALEX/PIE or something else?

Also, if you can provide a link to PTU documentation and/or code examples in other languages it would be great!

**Comment by snizzleorg** [comment]
I will get you one. PIE FRET. Any particular place where I can upload it? I can put it on dropbox. But thats probably not the best place.

**Comment by snizzleorg** [comment]
[Cy3+Cy5_diff_PIE-FRET.ptu.zip](https://github.com/Photon-HDF5/phconvert/files/231343/Cy3.Cy5_diff_PIE-FRET.ptu.zip)

here it is...


SAMPLE

RNA strands with a flexible tetraloop (labeled with Cy3) and the corresponding tetraloop receptor (Cy5) freely diffusing in buffer solution. Sample courtesy of Julie Fiore and David Nesbitt (JILA, Univ. of Colorado, Boulder, USA). For further details about the sample see:

C. D. Downey, J. L. Fiore, C. D. Stoddard, J. H. Hodak, D. J. Nesbitt and A. Pardi,
"Metal ion dependence, thermodynamics, and kinetics for intramolecular
docking of a GAAA tetraloop and receptor connected by a flexible linker,"
Biochem. 45, 3664-3674 (2006).  

ACQUISITION PARAMETERS

MicroTime 200 system, pulsed interleaved excitation with 532nm (PicoTA) and 640 nm, microscope objective 60x1,2 water immersion, detected fluorescence light spectrally separated and detected with 2 SPAD detectors

**Comment by tritemio** [comment]
@snizzleorg thanks a lot! 👍

**Comment by tritemio** [comment]
@snizzleorg, I looked and the C++ demo (the MATLAB demo seems incomplete to me) and I found that PTU files are quite more complex than PT3 and HT3.

Anyway I hacked a decoder for the file you posted:

https://gist.github.com/tritemio/734347586bc999f39f9ffe0ac5ba0e66

The tricky part was decoding the header which is made of fields (called "tags") of non-constant size.
Also, PTU files can contain t3records with different "formats". In the linked notebook, I just use the correct parameters to decode the example file you posted, which is a 'rtPicoHarpT3' file.

Another thing, I don't decode all possible data types in the header, in particular types "double" and "wide strings" (whatever it means) are missing (and that's ok for current file).

I don't plan to further work on this for the time being, so if anybody wants to take over is welcome. With the proper knowledge of PTU, extending this to the general case should be easy. Maybe an expert from PicoQuant can complete it and make a PR to phconvert?

**Comment by tritemio** [comment]
I updated the previous notebook adding the ability to decode the different data types in the header.

Still, it is limited to 'rtPicoHarpT3' files, but adding the other t3records formats should be easy.

**Comment by ncodina** [comment]
Hi,
Thanks very much @tritemio for writing a decoder for .PTU files. In my lab we would also like to convert .PTU to Photon-HDF5 file, but we have the type: "rtHydraHarp2T3".

We don't think the decoder written as it is is extracting the data correctly, but we don't know what changes should we make. Do you think you could help us convert the file?

Firstly, in [this section](https://render.githubusercontent.com/view/ipynb?enc_url=68747470733a2f2f7261772e67697468756275736572636f6e74656e742e636f6d2f676973742f74726974656d696f2f37333433343735383662633939396633396639666665306163356261306536362f7261772f306562326135333137303831376631366533613266353661333761303065343330633632616339382f4465636f64696e6725323050545525323066696c65732e6970796e62#Read-T3records) there is a line that looks like this:

```python
detectors, timestamps, nanotimes = \
    phc.pqreader.process_t3records(
        t3records, time_bit=16, dtime_bit=12, ch_bit=4, special_bit=False)
```

and here we are unsure what values to use for `time_bit`, `dtime_bit`, `ch_bit`, and `special_bit`. Could you let us know where we get these numbers from?

Secondly, we were wondering about the `detectors` array that is returned from `process_t3records`. In our instrument we have 2 detectors, yet this array returns 6 different values: 0,  1,  2,  4,  6, 15. Could you explain where these numbers come from?



I'm attaching an example file (dropbox link): "20161027_DM1_1nM_pH7_20MHz1.ptu"

https://www.dropbox.com/s/s8zzxcq7d2nfqe0/20161027_DM1_1nM_pH7_20MHz1.ptu.zip?dl=0

It is a 1 nM Fab protein in phosphate buffer at pH = 7. The fluorophores are located at a distance where FRET should happen. 
Acquisition details:
MicroTime 200 microscope. 
Single excitation laser at 485 nm, 20 MHz, about 100 μW at the back aperture.
Objective 60x/1.20W.
50-μm pinhole.
485/595 dual-band dichroic mirror, 585 dichroic, and 525/50 and 645/75 band-pass filters.  
2 SPAD detectors and HydraHarp 400.

**Comment by tritemio** [comment]
Hi @ncodina, you are welcome!

the line you pinpointed in the notebook is exactly the one that needs to be modified in order to correctly decode your file. To find the values to pass to the function I used the PicoQuant C example as reference (linked at the top of the notebook):

https://github.com/PicoQuant/PicoQuant-Time-Tagged-File-Format-Demos/blob/master/PTU/cc/ptudemo.cc

In their demo code they decode the rtHydraHarp2T3 record calling ` ProcessHHT3(TTTRRecord, 2)` ([link](https://github.com/PicoQuant/PicoQuant-Time-Tagged-File-Format-Demos/blob/master/PTU/cc/ptudemo.cc#L578)). The function is defined [here](https://github.com/PicoQuant/PicoQuant-Time-Tagged-File-Format-Demos/blob/master/PTU/cc/ptudemo.cc#L293).

Try to see if you can figure out the parameters and if not I'll look into it.

You should probably also ask PicoQuant if they provided some written documentation about PTU (they almost surely do).

**Comment by tritemio** [comment]
Just a clarification. The phconvert function `process_t3records` (in pqreader.py, [see implementation](https://github.com/Photon-HDF5/phconvert/blob/0.7.2/phconvert/pqreader.py#L328)) should basically do the same decoding as the `ProcessHHT3(TTTRRecord, 2)` call in the picoquant demo code.

**Comment by ncodina** [comment]
Hi @tritemio, thanks very much.

We changed the code with the new parameters for .PTU type "rtHydraHarp2T3":

```
detectors, timestamps, nanotimes = \
    phc.pqreader.process_t3records(
        t3records, time_bit=10, dtime_bit=15, ch_bit=6, special_bit=True)
```

and it extracted the 3 arrays: detectors, timestamps and nanotimes.

**Comment by tritemio** [comment]
Perfect, I'm glad it worked. I have updated the notebook which now supports "rtHydraHarp2T3" and "rtPicoHarpT3" records types.

I also added some example of how to retrieve metadata from the header (timestamps units, nanotimes units and acquisition duration).

The 127 detectors in your file are the overflow "timestamps" you can safely filter them out. The detectors 0 and 1 have the bulk of the data and are the "real" detectors. The other detectors (34,  50,  68) have only 1 or 2 counts, I don't know what they are.

Can you please send a very short file that we can use for testing phconvert?

Eventually we want PTU support directly in phconvert. Would you be able to send a Pull Request to integrate the PTU decoding in the phconvert library? If yes we can discuss further which functions to add.

**Comment by ncodina** [comment]
Thanks so much again!

Thanks for providing information about the additional detectors. We didn't know what the high counts in detector 127 were.

Yes, I can provide a shorter file. What is the maximum size? 10 MB?
I can provide a single laser excitation and a PIE sample.

And yes, we can generate a Pull Request that supports PTU decoding.

**Comment by tritemio** [comment]
If you can generate files of 100-200KB we can include them in the repository and add unit tests against them like it was done for Becker&Hickl SET/SPC files.

PS: if you want, you can always remove the photons which have `detectors == 127` before saving to Photon-HDF5.

**Comment by tritemio** [comment]
See here for a PR that added support for more Becker&Hickl files and added the unit tests:

https://github.com/Photon-HDF5/phconvert/pull/18

I was thinking of switching from `nose` to `pytest` as a testing framework because it simplifies writing tests (and can also run tests in `nose` format).

**Comment by tritemio** [comment]
I have copied the code from the notebook into the pqreader.py file in current phconvert master.  We still have to add tests and a function to load smFRET files from PTU files  in `loader.py` to load smFRET PTU files (similarly to `nsalex_ht3`). Please base your work on the master branch when you send a PR.

Note that the current master branch started implementing the new Photon-HDF5 format 0.5 (the current version of Photon-HDF5 is 0.4). Phconvert version 0.7.2 (latest release) still implements Photon-HDF5 0.4. The next phconvert version 0.8+ will come from the master branch and will implement Photon-HDF5 0.5.

**Comment by tritemio** [comment]
@ncodina, yes I went back to the PicoQuant demo code and I found that the overflow correction use a different logic (the number of overflows is stored in the timestamp).

Therefore, the implementation that I recently committed in phconvert master was wrong for  'rtHydraHarp2T3' records, resulting in apparently shorter measurement.

I fixed the bug in current master adding a new overflow correction function. Now you should be able to call `load_ptu(filename)` to load and decode the file.

Let me know if there are further problems. I plan to release a new version of phconvert with PTU support in a week or so.

**Comment by tritemio** [comment]
PS: to test current master version of phconvert, clone the repo locally (or download the zip), then enter the source folder and type:

```
pip install -e .
```

Double-check the exact version of phconvert you are using with:

```
import phconvert as phc
print('phconvert version: ' + phc.__version__)
```

**Comment by ncodina** [comment]
@tritemio thanks very very much!

I downloaded the most current version (phconvert version: 0.7+27.g5849b92) and the .h5 files generated are the correct acquisition duration. Thanks for realising and fixing the overflow correction.

**Comment by tritemio** [comment]
@ncodina, I'm glad it works for you too. I forgot to mention that current master has an additional fix in computing the PTU header size. Previously the last tag was not read resulting in 2 spurious counts with weird detector number (see the fix e883d895386c475b6960af8a46ec756a4cdc0c3a).

**Comment by tritemio** [comment]
@ncodina Can you provide a small PTU file can be used for testing, let's say a 30s acquisition. I will upload it to figshare and use as example and test file for phconvert.  For now I only have one PTU file you provided but it is almost 100MB and I would like to replace it with a smaller one for continuous integration.

Also if you are willing to, you can write an example notebook showing howto convert a smFRET measurement with lifetime from PTU to Photon-HDF5. You need to add a function in loader.py for this kind of measurement/format combination (take inspiration from the other functions there in).

**Comment by ncodina** [comment]
Yes, sorry, I realize I told you that I would send the sample. 
Here they are:
20s acquisition time
[161128_DM1_50pM_pH71.ptu.zip](https://github.com/Photon-HDF5/phconvert/files/616769/161128_DM1_50pM_pH71.ptu.zip)
15s acquisition time
[161128_DM1_50pM_pH73.ptu.zip](https://github.com/Photon-HDF5/phconvert/files/616770/161128_DM1_50pM_pH73.ptu.zip)
30s acquisition time
[161128_DM1_50pM_pH74.ptu.zip](https://github.com/Photon-HDF5/phconvert/files/616771/161128_DM1_50pM_pH74.ptu.zip)

The acquisition details are described in the file. It is a 50 pM protein labeled with one AF488 and one AF594. Let me know if you need to know something else.

I will write the notebook tomorrow if I can, or later in the week :).

**Comment by tritemio** [comment]
@ncodina, thanks for the files! For the notebook take your time. I'm going to release phconvert soon. The notebook can be added later.

**Comment by tritemio** [comment]
Last week I released phconvert 0.7.3 which includes PTU support with all the [unit tests]() in place. There was a delay (~2 days) in building the mac OS packages but now conda packages for 0.7.3 are available on conda-forge on all platforms.

Not all the possible t3records are still supported, but at this point adding new t3records is quite easy and will be done based on user request.

I'll close this issue for now. If and when there is interest in expanding the PTU support to new t3records  a new issue can be opened.

**Comment by ncodina** [comment]
@tritemio I just opened an issue because we are also interested in doing “smFRET-nsALEX” measurements. When we know how to write correctly this files I could do a notebook showing how to convert both types of measurements.
